In [2]:
import pandas as pd
import numpy as np

In [3]:
# loading the file:

column_names = ['mpg', 'cylinders', 'displacement', 'horsepower', 'weight', 'acceleration', \
                'model_year', 'origin', 'car_name']
df = pd.read_csv(r"C:\Users\adishree.b\Documents\auto-mpg.csv", delim_whitespace=True, names=column_names)
df.head()

mpg  cylinders  displacement horsepower  weight  acceleration  model_year  \
0  18.0          8         307.0      130.0  3504.0          12.0          70   
1  15.0          8         350.0      165.0  3693.0          11.5          70   
2  18.0          8         318.0      150.0  3436.0          11.0          70   
3  16.0          8         304.0      150.0  3433.0          12.0          70   
4  17.0          8         302.0      140.0  3449.0          10.5          70   

   origin                   car_name  
0       1  chevrolet chevelle malibu  
1       1          buick skylark 320  
2       1         plymouth satellite  
3       1              amc rebel sst  
4       1                ford torino

## Question 1: Create a new dataframe that contains all the values of the original dataframe but with no missing values. How many rows does this dataframe contain? (Hint: missing values denoted as ‘?’ in the horsepower column)

In [5]:
# let's check if there are any null values in the dataset

df.isnull().sum()

mpg             0
cylinders       0
displacement    0
horsepower      0
weight          0
acceleration    0
model_year      0
origin          0
car_name        0
dtype: int64

In [7]:
# We don't see any null values in the dataset, but as question suggests, let's check the horsepower column with "?" values

df[df['horsepower']=="?"]

mpg  cylinders  displacement horsepower  weight  acceleration  \
32   25.0          4          98.0          ?  2046.0          19.0   
126  21.0          6         200.0          ?  2875.0          17.0   
330  40.9          4          85.0          ?  1835.0          17.3   
336  23.6          4         140.0          ?  2905.0          14.3   
354  34.5          4         100.0          ?  2320.0          15.8   
374  23.0          4         151.0          ?  3035.0          20.5   

     model_year  origin              car_name  
32           71       1            ford pinto  
126          74       1         ford maverick  
330          80       2  renault lecar deluxe  
336          80       1    ford mustang cobra  
354          81       2           renault 18i  
374          82       1        amc concord dl

In [8]:
# Removing the rows where horsepower has '?' 
df = df[df.horsepower != '?']


In [9]:
# let's check the horsepower column with "?" values again:
df[df['horsepower']=="?"]

Empty DataFrame
Columns: [mpg, cylinders, displacement, horsepower, weight, acceleration, model_year, origin, car_name]
Index: []

#### We can see that all all the rows where horsepower has '?' has been removed.

## Question2:	Create a subset of this dataframe that contains 8 cylinders and weight greater than 3000. How many rows does this subset dataframe contain?

In [10]:
# let's check the shape of the dataset at without subsetting:
df.shape

(392, 9)

In [11]:
# Creating the subset of the dataframe that contains 8 cylinders and weight greater than 3000:

df_sub = df[(df['cylinders']== 8) & (df['weight']> 3000)]

df_sub.shape

(103, 9)

#### The subset dataframe contains 103 rows

## Question3: Which car name and model has the highest weight in this dataset? Which car name and model has the lowest weight?

In [17]:
df[['car_name', 'model_year']][df['weight']== df['weight'].max()]

car_name  model_year
44  pontiac safari (sw)          71

In [18]:
df[['car_name', 'model_year']][df['weight']== df['weight'].min()]

car_name  model_year
54  datsun 1200          71

#### The car name, model having highest weight in the dataset is "pontiac safari (sw), 71" and the one with lowest weight is "datsun 1200, 71" 


## Question 4:	Add another column to this dataset, called ‘model80s’, indicating whether the car model year is 80,81, or 82. Does this group models has higher or lower mean mpg than other models (< 80)?

In [22]:
import numpy as np

In [24]:
# Adding another column to this dataset, called ‘model80s’, indicating whether the car model year is 80,81, or 82:
# labeling the new column with '80's' and 'before_80s'

df['model80s'] = np.where((df['model_year']>=80),'80s','before_80s')

In [27]:
# comparing the mean mpg values between the '80s' and 'before_80's' types

df.groupby('model80s').mpg.mean()

model80s
80s           31.975294
before_80s    21.084365
Name: mpg, dtype: float64

#### We can see that the mean mpg value is higher in the 80s with 31.97 than in the ones before 80s with 21.08

## Question5.	Calculate the mean mpg and mean weight for each of the cylinders levels. Which cylinder level has the highest mean weight and which one has the highest mean mpg? 

In [32]:
# grouping the data by cylinder levels, getting the mean values for each and sorting in desc values of mean mpg

df.groupby('cylinders').weight.mean().sort_values(ascending = False)

cylinders
8    4114.718447
6    3202.120482
5    3103.333333
3    2398.500000
4    2305.110553
Name: weight, dtype: float64

####  The cylinder level having the highest weight is '8' with mean weight of 4114